File: frontend/lib/api/schemas.ts
import { z } from "zod";

export const healthSchema = z.object({
  status: z.string(),
  service: z.string(),
  nba_season: z.string(),
  database_configured: z.boolean(),
  raw_data_dir: z.string(),
  request_timeout_seconds: z.number(),
  checked_at: z.string(),
});

export const pageMetaSchema = z.object({
  limit: z.number(),
  offset: z.number(),
  total: z.number(),
  next_offset: z.number().nullable(),
  previous_offset: z.number().nullable(),
});

export const teamSchema = z.object({
  id: z.number(),
  nba_team_id: z.number(),
  abbreviation: z.string(),
  city: z.string(),
  name: z.string(),
  conference: z.string().nullable(),
  division: z.string().nullable(),
});

export const teamsResponseSchema = z.object({
  teams: z.array(teamSchema),
});

export const draftPickSchema = z.object({
  id: z.number(),
  draft_year: z.number(),
  round: z.number(),
  overall_pick: z.number(),
  player_name: z.string(),
  school_country: z.string(),
  team: teamSchema,
  transaction_note: z.string().nullable(),
});

export const draftResponseSchema = z.object({
  draft_year: z.number(),
  event_dates: z.string(),
  venue: z.string(),
  location: z.string(),
  total_picks: z.number(),
  picks: z.array(draftPickSchema),
  data_source: z.string(),
});

export const playoffPlayerBoxScoreSchema = z.object({
  player_id: z.number().nullable(),
  player_name: z.string(),
  points: z.number(),
  rebounds: z.number(),
  assists: z.number(),
  steals: z.number(),
  blocks: z.number(),
  turnovers: z.number(),
  field_goals_made: z.number(),
  field_goals_attempted: z.number(),
  three_pointers_made: z.number(),
  three_pointers_attempted: z.number(),
  free_throws_made: z.number(),
  free_throws_attempted: z.number(),
  plus_minus: z.number().nullable(),
});

export const playoffTeamBoxScoreSchema = z.object({
  team: teamSchema,
  points: z.number(),
  rebounds: z.number(),
  assists: z.number(),
  steals: z.number(),
  blocks: z.number(),
  turnovers: z.number(),
  field_goals_made: z.number(),
  field_goals_attempted: z.number(),
  three_pointers_made: z.number(),
  three_pointers_attempted: z.number(),
  free_throws_made: z.number(),
  free_throws_attempted: z.number(),
  players: z.array(playoffPlayerBoxScoreSchema),
});

export const playoffGameSchema = z.object({
  id: z.number(),
  game_number: z.number(),
  home_team: teamSchema,
  away_team: teamSchema,
  home_score: z.number(),
  away_score: z.number(),
  overtime: z.boolean(),
  data_note: z.string().nullable(),
  team_box_scores: z.array(playoffTeamBoxScoreSchema),
});

export const playoffRoundResponseSchema = z.object({
  season: z.string(),
  round: z.string(),
  stored_series: z.number(),
  stored_games: z.number(),
  expected_games: z.number(),
  coverage_complete: z.boolean(),
  data_source: z.string(),
  series: z.array(
    z.object({
      id: z.number(),
      conference: z.string(),
      winner: teamSchema,
      loser: teamSchema,
      winner_wins: z.number(),
      loser_wins: z.number(),
      games: z.array(playoffGameSchema),
    }),
  ),
});

export const playerListItemSchema = z.object({
  id: z.number(),
  nba_player_id: z.number(),
  slug: z.string(),
  full_name: z.string(),
  first_name: z.string().nullable(),
  last_name: z.string().nullable(),
  position: z.string().nullable(),
  jersey_number: z.string().nullable(),
  active: z.boolean(),
  team: teamSchema.nullable(),
});

export const playerDetailSchema = playerListItemSchema.extend({
  height: z.string().nullable(),
  weight_pounds: z.number().nullable(),
  birthdate: z.string().nullable(),
});

export const playersResponseSchema = z.object({
  items: z.array(playerListItemSchema),
  meta: pageMetaSchema,
});

export const seasonsResponseSchema = z.object({
  seasons: z.array(z.string()),
});

export const syncRunStatusSchema = z.object({
  id: z.number(),
  source: z.string(),
  status: z.string(),
  finished_at: z.string().nullable(),
  fetched_count: z.number(),
  inserted_count: z.number(),
  updated_count: z.number(),
  rejected_count: z.number(),
  error_message: z.string().nullable(),
});

export const dataStatusSchema = z.object({
  current_season: z.string(),
  player_count: z.number(),
  game_count: z.number(),
  player_game_record_count: z.number(),
  last_successful_sync: syncRunStatusSchema.nullable(),
  last_failed_sync: syncRunStatusSchema.nullable(),
});

export const playerSeasonSummarySchema = z.object({
  player_id: z.number(),
  nba_player_id: z.number(),
  season: z.string(),
  games_played: z.number(),
  minutes_per_game: z.number().nullable(),
  points_per_game: z.number().nullable(),
  rebounds_per_game: z.number().nullable(),
  assists_per_game: z.number().nullable(),
  turnovers_per_game: z.number().nullable(),
  plus_minus_per_game: z.number().nullable(),
  true_shooting_percentage: z.number().nullable(),
  true_shooting_source: z.string().nullable(),
  usage_rate: z.number().nullable(),
  points_per_36: z.number().nullable(),
  rebounds_per_36: z.number().nullable(),
  assists_per_36: z.number().nullable(),
  turnovers_per_36: z.number().nullable(),
  analytics_rebuilt_at: z.string().nullable(),
  analytics_warnings: z.array(z.record(z.string(), z.unknown())),
});

export const playerSeasonSummariesResponseSchema = z.object({
  items: z.array(playerSeasonSummarySchema),
});

export const gameLogItemSchema = z.object({
  id: z.number(),
  game_id: z.number(),
  nba_game_id: z.string(),
  game_date: z.string(),
  season: z.string(),
  matchup: z.string().nullable(),
  location: z.enum(["home", "away"]).nullable(),
  opponent: teamSchema.nullable(),
  result: z.enum(["W", "L"]).nullable(),
  days_since_previous_game: z.number().nullable(),
  minutes: z.number().nullable(),
  points: z.number(),
  rebounds: z.number(),
  assists: z.number(),
  steals: z.number(),
  blocks: z.number(),
  turnovers: z.number(),
  personal_fouls: z.number(),
  field_goals_made: z.number(),
  field_goals_attempted: z.number(),
  three_pointers_made: z.number(),
  three_pointers_attempted: z.number(),
  free_throws_made: z.number(),
  free_throws_attempted: z.number(),
  plus_minus: z.number().nullable(),
  calculated_true_shooting_percentage: z.number().nullable(),
});

export const playerGamesResponseSchema = z.object({
  items: z.array(gameLogItemSchema),
  meta: pageMetaSchema,
});

export const trendsResponseSchema = z.object({
  player_id: z.number(),
  nba_player_id: z.number(),
  season: z.string(),
  production_trend: z.string().nullable(),
  production_trend_value: z.number().nullable(),
  efficiency_trend: z.string().nullable(),
  efficiency_trend_value: z.number().nullable(),
  payload: z.record(z.string(), z.unknown()),
});

export const splitSchema = z.object({
  games: z.number(),
  minutes: z.number(),
  points: z.number(),
  rebounds: z.number(),
  assists: z.number(),
  true_shooting_percentage: z.number().nullable(),
});

export const splitsResponseSchema = z.object({
  player_id: z.number(),
  nba_player_id: z.number(),
  season: z.string(),
  home: splitSchema.nullable(),
  away: splitSchema.nullable(),
});

export const benchmarksResponseSchema = z.object({
  player_id: z.number(),
  nba_player_id: z.number(),
  season: z.string(),
  league_percentiles: z.record(z.string(), z.unknown()),
  position_percentiles: z.record(z.string(), z.unknown()),
  minutes_tier_percentiles: z.record(z.string(), z.unknown()),
  analytics_warnings: z.array(z.record(z.string(), z.unknown())),
});

export const similarPlayerSchema = z.object({
  player: playerListItemSchema,
  summary: playerSeasonSummarySchema,
  similarity_score: z.number(),
  shared_position: z.boolean(),
  minutes_difference: z.number().nullable(),
});

export const similarPlayersResponseSchema = z.object({
  player_id: z.number(),
  nba_player_id: z.number(),
  season: z.string(),
  players: z.array(similarPlayerSchema),
});

export const recentWindowSchema = z.object({
  window: z.number(),
  games: z.number(),
  points: z.number().nullable(),
  rebounds: z.number().nullable(),
  assists: z.number().nullable(),
  minutes: z.number().nullable(),
  true_shooting_percentage: z.number().nullable(),
  turnovers: z.number().nullable(),
  plus_minus: z.number().nullable(),
});

export const restSplitsSchema = z.object({
  first_game: splitSchema.nullable(),
  back_to_back: splitSchema.nullable(),
  one_day_rest: splitSchema.nullable(),
  two_or_more_days_rest: splitSchema.nullable(),
});

export const rollingTrendPointSchema = z.object({
  player_id: z.number(),
  nba_player_id: z.number(),
  game_id: z.number(),
  game_date: z.string(),
  points_rolling_5: z.number().nullable(),
  points_rolling_10: z.number().nullable(),
  assists_rolling_5: z.number().nullable(),
  assists_rolling_10: z.number().nullable(),
  true_shooting_percentage_rolling_5: z.number().nullable(),
  true_shooting_percentage_rolling_10: z.number().nullable(),
});

export const compareMetricSchema = z.object({
  key: z.string(),
  label: z.string(),
  category: z.string(),
  higher_is_better: z.boolean(),
  player_a_value: z.number().nullable(),
  player_b_value: z.number().nullable(),
  winner: z.enum(["player_a", "player_b", "tie"]).nullable(),
});

export const comparePlayerSchema = z.object({
  player: playerListItemSchema,
  summary: playerSeasonSummarySchema,
  trends: trendsResponseSchema,
  splits: splitsResponseSchema,
  benchmarks: benchmarksResponseSchema,
  rest_splits: restSplitsSchema,
  recent_5: recentWindowSchema,
  recent_10: recentWindowSchema,
  rolling_trend: z.array(rollingTrendPointSchema),
});

export const compareResponseSchema = z.object({
  season: z.string(),
  player_a: comparePlayerSchema,
  player_b: comparePlayerSchema,
  position_match: z.boolean(),
  position_context: z.string(),
  category_winners: z.array(compareMetricSchema),
});

export const reportSentenceSchema = z.object({
  text: z.string(),
  fields: z.array(z.string()),
});

export const reportSectionSchema = z.object({
  key: z.string(),
  title: z.string(),
  sentences: z.array(reportSentenceSchema),
});

export const playerReportResponseSchema = z.object({
  player: playerListItemSchema,
  season: z.string(),
  summary: playerSeasonSummarySchema,
  trends: trendsResponseSchema,
  splits: splitsResponseSchema,
  benchmarks: benchmarksResponseSchema,
  sample_size_warnings: z.array(z.record(z.string(), z.unknown())),
  sections: z.array(reportSectionSchema),
});

export const assistantSourceSchema = z.object({
  tool: z.string(),
  arguments: z.record(z.string(), z.unknown()),
});

export const assistantChatResponseSchema = z.object({
  session_id: z.string(),
  message: z.string(),
  mode: z.string(),
  sources: z.array(assistantSourceSchema),
  context: z.object({
    last_entity_type: z.string().nullable(),
    last_entity_name: z.string().nullable(),
  }),
  data: z.record(z.string(), z.unknown()).nullable(),
});

export const awardCandidateSchema = z.object({
  rank: z.number(),
  entity_id: z.number(),
  name: z.string(),
  team: z.string().nullable(),
  position: z.string().nullable(),
  probability: z.number(),
  score: z.number(),
  features: z.record(z.string(), z.number()),
  feature_attributions: z.array(
    z.object({
      feature: z.string(),
      raw_value: z.number(),
      normalized_value: z.number(),
      weight: z.number(),
      contribution: z.number(),
    }),
  ),
  warnings: z.array(z.string()),
});

export const awardPredictionSchema = z.object({
  prediction_id: z.number().nullable(),
  award_type: z.string(),
  target_season: z.string(),
  source_season: z.string(),
  model_version: z.string(),
  candidates: z.array(awardCandidateSchema),
  warnings: z.array(z.string()),
  methodology: z.string(),
});

export const projectedTeamSchema = z.object({
  rank: z.number(),
  conference_rank: z.number().nullable(),
  team_id: z.number(),
  nba_team_id: z.number(),
  team_name: z.string(),
  abbreviation: z.string(),
  conference: z.string().nullable(),
  division: z.string().nullable(),
  projected_wins: z.number(),
  projected_losses: z.number(),
  projected_point_differential: z.number(),
  confidence: z.number(),
  factors: z.array(
    z.object({
      feature: z.string(),
      raw_value: z.number(),
      contribution: z.number(),
    }),
  ),
  warnings: z.array(z.string()),
});

export const standingsPredictionSchema = z.object({
  prediction_id: z.number().nullable(),
  target_season: z.string(),
  source_season: z.string(),
  conference: z.string().nullable(),
  model_version: z.string(),
  teams: z.array(projectedTeamSchema),
  warnings: z.array(z.string()),
  methodology: z.string(),
});

export const predictionLogSchema = z.object({
  predictions: z.array(
    z.object({
      id: z.number(),
      created_at: z.string(),
      target_season: z.string(),
      prediction_type: z.string(),
      model_version: z.string(),
      notes: z.string().nullable(),
      payload: z.record(z.string(), z.unknown()),
    }),
  ),
});

export type BackendHealth = z.infer<typeof healthSchema>;
export type AssistantChatResponse = z.infer<typeof assistantChatResponseSchema>;
export type AssistantSource = z.infer<typeof assistantSourceSchema>;
export type AwardCandidate = z.infer<typeof awardCandidateSchema>;
export type AwardPrediction = z.infer<typeof awardPredictionSchema>;
export type CompareMetric = z.infer<typeof compareMetricSchema>;
export type ComparePlayer = z.infer<typeof comparePlayerSchema>;
export type CompareResponse = z.infer<typeof compareResponseSchema>;
export type DataStatus = z.infer<typeof dataStatusSchema>;
export type DraftPick = z.infer<typeof draftPickSchema>;
export type DraftResponse = z.infer<typeof draftResponseSchema>;
export type PlayoffRoundResponse = z.infer<typeof playoffRoundResponseSchema>;
export type GameLogItem = z.infer<typeof gameLogItemSchema>;
export type PageMeta = z.infer<typeof pageMetaSchema>;
export type PlayerBenchmarks = z.infer<typeof benchmarksResponseSchema>;
export type PlayerDetail = z.infer<typeof playerDetailSchema>;
export type PlayerGamesResponse = z.infer<typeof playerGamesResponseSchema>;
export type PlayerReportResponse = z.infer<typeof playerReportResponseSchema>;
export type PlayerListItem = z.infer<typeof playerListItemSchema>;
export type PlayersResponse = z.infer<typeof playersResponseSchema>;
export type PlayerSeasonSummary = z.infer<typeof playerSeasonSummarySchema>;
export type PlayerSeasonSummariesResponse = z.infer<typeof playerSeasonSummariesResponseSchema>;
export type RecentWindow = z.infer<typeof recentWindowSchema>;
export type ReportSection = z.infer<typeof reportSectionSchema>;
export type RestSplits = z.infer<typeof restSplitsSchema>;
export type RollingTrendPoint = z.infer<typeof rollingTrendPointSchema>;
export type SimilarPlayer = z.infer<typeof similarPlayerSchema>;
export type SimilarPlayersResponse = z.infer<typeof similarPlayersResponseSchema>;
export type PlayerSplits = z.infer<typeof splitsResponseSchema>;
export type PlayerTrends = z.infer<typeof trendsResponseSchema>;
export type SeasonsResponse = z.infer<typeof seasonsResponseSchema>;
export type SplitSummary = z.infer<typeof splitSchema>;
export type Team = z.infer<typeof teamSchema>;
export type TeamsResponse = z.infer<typeof teamsResponseSchema>;
export type PredictionLog = z.infer<typeof predictionLogSchema>;
export type ProjectedTeam = z.infer<typeof projectedTeamSchema>;
export type StandingsPrediction = z.infer<typeof standingsPredictionSchema>;

File: backend/app/api/schemas.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ApiError(BaseModel):
    code: str
    message: str


class ErrorResponse(BaseModel):
    error: ApiError


class ApiMetadata(BaseModel):
    season: str
    dataFreshness: datetime | None
    source: str = "database"


class PageMeta(BaseModel):
    limit: int
    offset: int
    total: int
    next_offset: int | None
    previous_offset: int | None
    season: str | None = None
    dataFreshness: datetime | None = None
    source: str = "database"


class TeamResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nba_team_id: int
    abbreviation: str
    city: str
    name: str
    conference: str | None
    division: str | None


class PlayerListItem(BaseModel):
    id: int
    nba_player_id: int
    slug: str
    full_name: str
    first_name: str | None
    last_name: str | None
    position: str | None
    jersey_number: str | None
    active: bool
    team: TeamResponse | None


class PlayerDetail(PlayerListItem):
    height: str | None
    weight_pounds: int | None
    birthdate: date | None
    meta: ApiMetadata | None = None


class PaginatedPlayersResponse(BaseModel):
    items: list[PlayerListItem]
    meta: PageMeta


class SeasonItemResponse(BaseModel):
    season: str
    displayName: str
    isCompleted: bool
    isCurrent: bool
    supportsPredictions: bool
    supportsJordanPredictions: bool


class SeasonsResponse(BaseModel):
    seasons: list[str]
    data: list[SeasonItemResponse] = Field(default_factory=list)


class PositionsResponse(BaseModel):
    positions: list[str]


class TeamsResponse(BaseModel):
    teams: list[TeamResponse]
    meta: ApiMetadata | None = None


class TeamSeasonSummaryApiResponse(BaseModel):
    team: TeamResponse
    season: str
    conference: str | None
    division: str | None
    wins: int | None
    losses: int | None
    win_pct: float | None
    conference_rank: int | None
    division_rank: int | None
    points_per_game: float | None
    points_allowed_per_game: float | None
    net_rating: float | None
    offensive_rating: float | None
    defensive_rating: float | None
    pace: float | None
    playoff_result: str | None
    meta: ApiMetadata


class RosterMemberResponse(BaseModel):
    player: PlayerListItem
    jersey_number: str | None
    position: str | None
    roster_status: str
    is_projected_starter: bool
    depth_order: int | None


class TeamRosterResponse(BaseModel):
    team: TeamResponse
    members: list[RosterMemberResponse]
    meta: ApiMetadata


class TeamDetailResponse(BaseModel):
    team: TeamResponse
    summary: TeamSeasonSummaryApiResponse | None
    meta: ApiMetadata


class StandingRowResponse(BaseModel):
    team: TeamResponse
    conference: str
    rank: int
    wins: int
    losses: int
    win_pct: float
    games_back: float | None
    conference_record: str | None
    division_record: str | None
    home_record: str | None
    away_record: str | None
    last_10: str | None
    streak: str | None


class StandingsResponse(BaseModel):
    standings: list[StandingRowResponse]
    meta: ApiMetadata


class DraftPickResponse(BaseModel):
    id: int
    draft_year: int
    round: int
    overall_pick: int
    player_name: str
    school_country: str
    team: TeamResponse
    transaction_note: str | None


class DraftResponse(BaseModel):
    draft_year: int
    event_dates: str
    venue: str
    location: str
    total_picks: int
    picks: list[DraftPickResponse]
    data_source: str


class PlayoffPlayerBoxScoreResponse(BaseModel):
    player_id: int | None
    player_name: str
    points: int
    rebounds: int
    assists: int
    steals: int
    blocks: int
    turnovers: int
    field_goals_made: int
    field_goals_attempted: int
    three_pointers_made: int
    three_pointers_attempted: int
    free_throws_made: int
    free_throws_attempted: int
    plus_minus: int | None


class PlayoffTeamBoxScoreResponse(BaseModel):
    team: TeamResponse
    points: int
    rebounds: int
    assists: int
    steals: int
    blocks: int
    turnovers: int
    field_goals_made: int
    field_goals_attempted: int
    three_pointers_made: int
    three_pointers_attempted: int
    free_throws_made: int
    free_throws_attempted: int
    players: list[PlayoffPlayerBoxScoreResponse]


class PlayoffGameResponse(BaseModel):
    id: int
    game_number: int
    home_team: TeamResponse
    away_team: TeamResponse
    home_score: int
    away_score: int
    overtime: bool
    data_note: str | None
    team_box_scores: list[PlayoffTeamBoxScoreResponse]


class PlayoffSeriesResponse(BaseModel):
    id: int
    conference: str
    winner: TeamResponse
    loser: TeamResponse
    winner_wins: int
    loser_wins: int
    games: list[PlayoffGameResponse]


class PlayoffRoundResponse(BaseModel):
    season: str
    round: str
    stored_series: int
    stored_games: int
    expected_games: int
    coverage_complete: bool
    data_source: str
    series: list[PlayoffSeriesResponse]


class PlayerSeasonSummaryResponse(BaseModel):
    player_id: int
    nba_player_id: int
    season: str
    games_played: int
    minutes_per_game: float | None
    points_per_game: float | None
    rebounds_per_game: float | None
    assists_per_game: float | None
    turnovers_per_game: float | None
    plus_minus_per_game: float | None
    true_shooting_percentage: float | None = Field(
        description="Calculated true-shooting percentage when true_shooting_source is set."
    )
    true_shooting_source: str | None
    usage_rate: float | None
    points_per_36: float | None
    rebounds_per_36: float | None
    assists_per_36: float | None
    turnovers_per_36: float | None
    analytics_rebuilt_at: datetime | None
    analytics_warnings: list[dict[str, Any]]
    meta: ApiMetadata | None = None


class PlayerSeasonSummariesResponse(BaseModel):
    items: list[PlayerSeasonSummaryResponse]


class GameLogItem(BaseModel):
    id: int
    game_id: int
    nba_game_id: str
    game_date: date
    season: str
    matchup: str | None
    location: str | None
    opponent: TeamResponse | None
    result: str | None
    days_since_previous_game: int | None
    minutes: float | None
    points: int
    rebounds: int
    assists: int
    steals: int
    blocks: int
    turnovers: int
    personal_fouls: int
    field_goals_made: int
    field_goals_attempted: int
    three_pointers_made: int
    three_pointers_attempted: int
    free_throws_made: int
    free_throws_attempted: int
    plus_minus: float | None
    calculated_true_shooting_percentage: float | None


class PaginatedGameLogsResponse(BaseModel):
    items: list[GameLogItem]
    meta: PageMeta


class TrendsResponse(BaseModel):
    player_id: int
    nba_player_id: int
    season: str
    production_trend: str | None
    production_trend_value: float | None
    efficiency_trend: str | None
    efficiency_trend_value: float | None
    payload: dict[str, Any]
    meta: ApiMetadata | None = None


class SplitResponse(BaseModel):
    games: int
    minutes: float
    points: int
    rebounds: int
    assists: int
    true_shooting_percentage: float | None


class PlayerSplitsResponse(BaseModel):
    player_id: int
    nba_player_id: int
    season: str
    home: SplitResponse | None
    away: SplitResponse | None
    meta: ApiMetadata | None = None


class BenchmarksResponse(BaseModel):
    player_id: int
    nba_player_id: int
    season: str
    league_percentiles: dict[str, Any]
    position_percentiles: dict[str, Any]
    minutes_tier_percentiles: dict[str, Any]
    analytics_warnings: list[dict[str, Any]]
    meta: ApiMetadata | None = None


class SimilarFeatureComparison(BaseModel):
    feature: str
    label: str
    unit: str
    player_value: float | None
    candidate_value: float | None
    difference: float | None


class SimilarPlayerResponse(BaseModel):
    player: PlayerListItem
    summary: PlayerSeasonSummaryResponse
    similarity_score: float = Field(
        ge=0,
        le=100,
        description=(
            "Cosine similarity of the standardized statistical feature vectors, mapped to 0-100. "
            "Higher means a more similar statistical profile, not an identical play style."
        ),
    )
    shared_position: bool
    minutes_difference: float | None
    shared_strengths: list[str] = Field(
        default_factory=list,
        description="Features where both players rate well above league average.",
    )
    largest_differences: list[str] = Field(
        default_factory=list,
        description="Features where the two standardized profiles diverge the most.",
    )
    feature_comparisons: list[SimilarFeatureComparison] = Field(default_factory=list)


class SimilarPlayersResponse(BaseModel):
    player_id: int
    nba_player_id: int
    season: str
    players: list[SimilarPlayerResponse]
    meta: ApiMetadata | None = None


class RecentWindowResponse(BaseModel):
    window: int
    games: int
    points: float | None
    rebounds: float | None
    assists: float | None
    minutes: float | None
    true_shooting_percentage: float | None
    turnovers: float | None
    plus_minus: float | None


class RestSplitsResponse(BaseModel):
    first_game: SplitResponse | None
    back_to_back: SplitResponse | None
    one_day_rest: SplitResponse | None
    two_or_more_days_rest: SplitResponse | None


class RollingTrendPointResponse(BaseModel):
    player_id: int
    nba_player_id: int
    game_id: int
    game_date: date
    points_rolling_5: float | None
    points_rolling_10: float | None
    assists_rolling_5: float | None
    assists_rolling_10: float | None
    true_shooting_percentage_rolling_5: float | None
    true_shooting_percentage_rolling_10: float | None


class CompareMetricResponse(BaseModel):
    key: str
    label: str
    category: str
    higher_is_better: bool
    player_a_value: float | None
    player_b_value: float | None
    winner: str | None = Field(description="player_a, player_b, tie, or null when unavailable.")


class ComparePlayerResponse(BaseModel):
    player: PlayerListItem
    summary: PlayerSeasonSummaryResponse
    trends: TrendsResponse
    splits: PlayerSplitsResponse
    benchmarks: BenchmarksResponse
    rest_splits: RestSplitsResponse
    recent_5: RecentWindowResponse
    recent_10: RecentWindowResponse
    rolling_trend: list[RollingTrendPointResponse]


class CompareResponse(BaseModel):
    season: str
    player_a: ComparePlayerResponse
    player_b: ComparePlayerResponse
    position_match: bool
    position_context: str
    category_winners: list[CompareMetricResponse]
    meta: ApiMetadata | None = None


class ReportSentenceResponse(BaseModel):
    text: str
    fields: list[str]


class ReportSectionResponse(BaseModel):
    key: str
    title: str
    sentences: list[ReportSentenceResponse]


class PlayerReportResponse(BaseModel):
    player: PlayerListItem
    season: str
    summary: PlayerSeasonSummaryResponse
    trends: TrendsResponse
    splits: PlayerSplitsResponse
    benchmarks: BenchmarksResponse
    sample_size_warnings: list[dict[str, Any]]
    sections: list[ReportSectionResponse]
    meta: ApiMetadata | None = None
